_drafts/arm-cortex-m-exceptions-and-nvic.md

```diff
@@ -0,0 +1,696 @@
+
+- **Exception Number** This number gives the offset within the **Vector Table** where the address of
+  the routine for the exception can be found. The routine is usually referred to as the **Exception
+  Handler** or **Interrupt Service Routine** (**ISR**) and is a pointer to the function which runs when
+  the exception is triggered. The ARM hardware will automatically look up this function pointer
+  when an exception is triggered and start executing the code.
+- **Priority Level** / **Priority Number** - Each exception has a priority associated with it. For most
+  exceptions this number is configurable. Counter-intuitively, the lower the priority number, the
+  higher the precedence the exception has. So for example if an exception of priority level 2 and
+  level 1 occur at the same time, the level 1 exception will be run first. When we say an exception
+  has the "highest priority", it will have the lowest **Priority Number**. If two exceptions have the _same_ priority **Priority Number**, the exception with the lowest **Exception Number** will run first.
+- **Synchronous** or **Asynchronous** - As the name implies, some exceptions will fire immediately
+  after an instruction is executed (i.e `SVCall`). These exceptions are referred to as
+  _synchronous_. Exceptions that do not fire immediately after a particular code path is executed are
+  referred to as _asynchronous_.
+
+An exception can be in one of several states:
+
+- **Pending** - The MCU has detected the exception and scheduled it but has not yet invoked the handler for it
+- **Active** - The MCU has started to run the exception handler but not yet finished executing it. It's possible for the exception to have been "pre-empted" by a higher priority handler
+- **Pending & Active** - Only possible for asynchronous exceptions, this basically means the exception was detected by the MCU again while processing an earlier detected version of the same exception.
+- **Inactive** - The exception is neither pending or active
+
+Some exceptions can be selectively Enabled or Disabled.
+
+> NOTE: Even while an exception is disabled, it can still reach the pending state. Upon being
+> enabled it will then transition to active. It's generally a good idea to clear any pending exceptions for an interrupt before enabling them.
+
+Let's explore the different types of exceptions available on ARM Cortex-M MCUs:
+
+### Built in Exceptions
+
+These are exceptions that are part of every ARM Cortex-M core. The ARM Cortex-M specifications
+reserve **Exception Number**s **1**-**15**, inclusive, for these.
+
+> NOTE: Recall that the **Exception Number** maps to an offset within the **Vector Table**. Index 0 of
+> the Vector Table holds the reset value of the Main stack pointer. The rest of the **Vector Table**
+> holds **Exception Handler** pointers which is why the **Exception Number**s start at 1
```

Comment:
They could have just done -1 to get the table index? Perhaps a more plausible reason so that zero can be used to indicate "no interrupt" (i.e. for `VECTACTIVE`, `VECTPENDING` etc)?

Reply:
heh, good observation point, I think I derived this statement from the following:

<img width="732" alt="image" src="https://user-images.githubusercontent.com/7494333/64207754-44a50a00-ce6b-11e9-8cc8-6389afb96c48.png">

I'll rework the wording slightly here, thanks!